### DISCUSSION TITLE
Different LUA variable names but same LUA structure

### DISCUSSION THREAD

**Discussion by DiagProf** [neolithos/neolua#150] [Q&A] [answered]
Hello, I hope I can ask this here. 
I want to use Lua to simulate the responses from ECUs. 
And I'm just struggling a bit with getting into neolua and also the lua topic.

I have Lua files that sometimes look like this

ABS = {
	RequestId = 0x713,
	ResponseId = 0x77D,
	Raw = {
		["3E 00"] = "7E 00",
		["22 F1 9E"] = "62 F1 9E 45 56 5F 45 53 50 39 42 4F 53 43 48 43 4D 50 00",
    }
}

or like this 

EMS = {
	RequestId = 0x70C,
	ResponseId = 0x776,
	Raw = {
		["3E 00"] = "7E 00",
		["22 F1 9E"] = "62 F1 9E 45 56 5F 53 4D 4C 53 4B 4C 4F 41 55 37 33 36 00",
		["22 F1 A2"] = "62 F1 A2 30 30 37 30 31 32",
		["10 03"] = "50 03 00 32 01 F4",
		["2E F1 98 00 00 00 00 00 2E"] = "6E F1 98",
		["2E F1 99 21 11 05"] = "6E F1 99",
    }
}

What is the best way to access this if you don't know what the top lua variable is named? (in these examples "ABS" and "EMS")

And what would be the correct way to access Raw?
As an example if e.g. the Request is ["3E 00"] the appropriate answer would be "7E 00"

that's my attempts so far

    internal class Program
    {
        private static LuaChunk script;
        private static dynamic _environment;

        private static void Main(string[] args)
        {
            using (var lua = new Lua()) // Create the Lua script engine
            {
                var path = "../../../BV_SteerColumElectUDS.lua";
                _environment = lua.CreateEnvironment<LuaGlobal>(); // Create a environment
                script = lua.CompileChunk(File.ReadAllText(Path.GetFullPath(path)), Path.GetFileName(path), new LuaCompileOptions { DebugEngine = LuaExceptionDebugger.Default });
                _environment.dochunk(script);
                Console.WriteLine(_environment.EMS.RequestId);
                //var raw = _environment.EMS.Raw;
            }
        }
    }
	
some tips would help me a lot.
thank you

**Comment by neolithos** [comment, marked as answer]
In Lua all data is stored in a table(s), which is a combination of an array and dictionary.

There are multiple ways to access.

C#:
- If you know the name, or it is fixed: Use the dynamic syntax. This will create on the first a access a little shortcut.
  `var raw = (LuaTable)_environment.EMS.Raw;` with the cast you can switch between dynamic/static.
- If you do not know the name or index: Use `table.ArrayList` or `table.Members`

Lua:
- dynamic, like lua, use `local raw = EMS.Raw`
- static, use ipairs and mpairs. Or the NeoLua specials rawarray() in combination with `foreach`.

**Reply by DiagProf** [reply]
Hello, sorry for the very very late reply. 
The project is such a free time idea and I haven't had the time for a long time.

With the useful tips I'm now trying to find the entry point. 
Would that be okay? 
Or too complicated?

`using ( Lua l = new Lua() )
            {
                LuaGlobal environment = l.CreateEnvironment<LuaGlobal>();
                LuaChunk chunk = null;
                try
                {
                    chunk = l.CompileChunk($"../../../Simulate_ECU.lua", new LuaCompileOptions() { DebugEngine = LuaExceptionDebugger.Default });
                }
                catch ( LuaParseException e )
                {
                    Debug.Fail("[LUA ERROR] " + e.Message + "\n" + e.FileName + " (line " + e.Line + ")");
                }

                try
                {
                    var result = environment.DoChunk(chunk);
                }
                catch ( Exception e )
                {
                    Console.WriteLine("Expception: {0}", e.Message);
                    var d = LuaExceptionData.GetData(e); // get stack trace
                    Console.WriteLine("StackTrace: {0}", d.FormatStackTrace(0, false));
                }

                var entryPointName = "";
                foreach ( var item in environment.Members )
                {
                    if ( item.Value is not LuaTable )
                    {
                        continue;
                    }

                    Console.WriteLine($"{item.Key} , {item.Value}");
                    var table = (LuaTable)environment[item.Key];
                    foreach ( var item2 in table.Members )
                    {
                        if ( item2.Key.Equals("Raw") )
                        {
                            entryPointName = item.Key;
                            Console.WriteLine($"Name of Entry point: {item.Key}");
                        }
                    }
                }

                if ( !string.IsNullOrWhiteSpace(entryPointName) )
                {
                    var lt = (LuaTable)((LuaTable)environment[entryPointName]).Members["Raw"];
                    foreach ( var testerReq in lt )
                    {
                        Console.WriteLine($"Req: {testerReq.Key} ->  Res: {testerReq.Value}");
                    }
                }
            }`
thanks

**Reply by neolithos** [reply]
- `if (item.Value is LuaTable table)` is shorter an easier.
- Take a look into `table.GetOptionalValue`, that might help

**Reply by DiagProf** [reply]
Thanks a lot
